src/api/informer.rs

```diff
@@ -126,43 +118,72 @@ impl<K> Informer<K> where
         self
     }
 
-
     /// Run a single watch poll
     ///
     /// If this returns an error, it resets the resourceVersion.
     /// This is meant to be run continually and events are meant to be handled between.
-    /// If handling all the events is too time consuming, you probably need a queue.
-    pub async fn poll(&self) -> Result<()> {
+    /// poll returns a Stream so events can be handled asynchronously
+    pub async fn poll(&self) -> Result<impl Stream<Item = Result<WatchEvent<K>>>> {
         trace!("Watching {:?}", self.resource);
-        match self.single_watch().await {
-            Ok((events, newver)) => {
-                *self.version.write().unwrap() = newver;
-                for e in events {
-                    self.events.write().unwrap().push_back(e);
-                }
-            },
+
+        // Create our watch request
+        let req = self.resource.watch(&self.params, &self.version())?;
+
+        // Clone our version so we can move it into the Stream handling
+        // and avoid a 'static lifetime on self
+        let version = self.version.clone();
+
+        // Fetch an initial version as well to use for if we encounter
+        // a desync issue
+        let initial_version = self.get_resource_version().await?;
+
+        // Attempt to fetch our stream
+        let stream = self.client.request_events::<WatchEvent<K>>(req).await;
+
+        match stream {
+            Ok(events) => {
+                // Add a map stage to the stream which will update our version
+                // based on each incoming event
+                Ok(events.map(move |event| {
+                    // Check if we need to update our version based on the incoming events
+                    let new_version = match &event {
+                        Ok(WatchEvent::Added(o))
+                        | Ok(WatchEvent::Modified(o))
+                        | Ok(WatchEvent::Deleted(o)) => o.meta().resourceVersion.clone(),
+                        _ => None,
+                    };
+
+                    // If we ran into an WatchError, log it and then reset our version
+                    if let Ok(WatchEvent::Error(e)) = &event {
```

Comment:
As far as the resetting ([comment](https://github.com/clux/kube-rs/pull/92#issuecomment-558302893)), looking at the `reset` function all it seems to do is get us back to our initial resource version, so I've done something similar here. Fetch it before the stream and then revert to it on any error.

Reply:
Ah ignore me I was misunderstanding the code, in non-stream version the `reset` forces a wait before attempting to repoll to help, while the current streaming implementation doesn't. I'll see what I can do to that end, thanks!